File: 16/main.py
from dataclasses import dataclass
from functools import lru_cache, partial
from itertools import permutations
from math import factorial
from multiprocessing import SimpleQueue
from queue import LifoQueue, PriorityQueue
from random import random
import re
from typing import List

@dataclass
class Valve:
    name: str
    flow: int
    to: ["Valve"]

    def __repr__(self):
        return f":{self.name}" # ({self.flow}) -> [{','.join(t.name for t in self.to)}]"
    def __hash__(self):
        return hash(self.name)
    def __eq__(self, other: object) -> bool:
        return self.name == other.name

def load(fname):
    valves = {}
    with open(fname) as f:
        for line in f:
            m = re.match("Valve ([A-Z][A-Z]) has flow rate=(\d+); tunnels? leads? to valves? ([A-Z, ]+)", line.strip())
            name,flow,to = m.groups()
            valves[name] = Valve(name,int(flow),to.split(', '))
    for valve in valves.values():
        valve.to = [valves[t] for t in valve.to]
    return valves

@lru_cache
def min_distance(s,e):
    q = SimpleQueue() # BFS
    q.put([s]) 
    visited = set([])
    while not q.empty():
        p = q.get()
        n = p[-1]
        visited.add(n)
        #print(p)
        if n == e:
            return len(p)-1
        for v in n.to:
            if v not in visited:
                q.put(p+[v])


def compute_flow(nodes):
    nodes = list(nodes)
    time = 30
    rate = 0
    total = 0
    for prev, node in zip(nodes,nodes[1:]):
        dt = min_distance(prev,node)+1
        if dt > time:
            break
        total += dt*rate
        rate += node.flow
        time -= dt
        #print(node, dt, rate, total, time)
    
    total += time*rate
    return total

def find_max_reward_path(start):
    valves_with_flow = [v for v in valves.values() if v.flow > 0]
    best_flow = 0
    total = factorial(len(valves_with_flow))
    from tqdm import tqdm
    for rest in tqdm(permutations(valves_with_flow), total=total):
        p = [start, *rest]
        flow = compute_flow(p)
        if flow > best_flow:
            best_flow = flow
            print(best_flow, p)
    return best_flow        

    #return max(map(compute_flow,([start,*rest] for rest in permutations(valves_with_flow))))


valves = load("test")
print(find_max_reward_path(valves["AA"]))

#test_path = get_valves(c+c for c in "ADBJHEC")
#test_path_2 = get_valves(c+c for c in "ABJDHFEC")
#print(compute_flow(test_path))

File: 16/main.ts
import { readFileSync } from 'fs';

interface Valve {
    name: string;
    flow: number;
    to: string[];
}

interface State {
    valve: string;
    time: number;
    open: Set<string>;
    pressure: number;
}

interface VisitedState {
    open: Set<string>;
    current: string;
}

let parseLine = function(line: string): Valve {
    const re = /Valve ([A-Z][A-Z]) has flow rate=(\d+); tunnels? leads? to valves? ([A-Z, ]+)/
    let [_, name, flow, to, ...rest] = re.exec(line)!;
    return {name, flow: +flow, to: to.split(', ')}
} 

let min_distance = function(s: string, e: string) : number {
    let q = [[s]]; 
    let visited : Set<string> = new Set();

    while (q.length > 0) {
        let p = q.shift()!; // BFS, it only gives shortest path for unweighted graphs
        let n = p[p.length -1];
        visited.add(n);
        if (n === e) { 
            return p.length -1;
        }
        neighbors(n).filter(v => !visited.has(v)).forEach(v => q.push(p.concat(v)));
    }
    return -1;
}

let neighbors = function(s: string): string[] {
    return valves[s].to
}

let union = function<T>(a: Set<T>, b: Set<T>) : Set<T> { return new Set([...a,...b]) };

let find_max_reward = function(time: number) : number {
     
    let valves_with_flow = Object.entries(valves).filter(([k,v]) => v.flow > 0).map(([k,v]) => k)
    let start : State = {valve: "AA", time, open: new Set(), pressure:0}
    let q = [start];
    
    //solutions is a map of open nodes to the pressure they can release for time iterations
    let solutions : Map<Set<string>, number> = new Map();

    let visited : Map<VisitedState, number> = new Map();
    let it = 0;
    
    while (q.length > 0) {
        it += 1
        if (it % 1_000 == 0) 
            console.log(it, solutions.size, Math.max(...solutions.values()));
   
        let s = q.pop()!;
        
        if (s.pressure > (solutions.get(s.open) ?? 0)) 
            solutions.set(s.open, s.pressure);
        
        valves_with_flow.forEach(v => {
            if (s.open.has(v)) return;
            //if (valves[v].flow == 0) return;
            let d = min_distance(s.valve, v);
            let new_time = s.time - d - 1;

            let sn : State = {
                valve: v,
                time: new_time,
                pressure: s.pressure + new_time * valves[v].flow,
                open: union(s.open, new Set([v]))
            };
            
            let vs : VisitedState = {open: sn.open, current: sn.valve}
            console.log(visited);
            if (sn.pressure > (visited.get(vs) ?? 0)) {
                visited.set(vs, sn.pressure);
                q.push(sn);
            }
        })
    } 
    console.log(it, "iterations");
    return Math.max(...solutions.values());

}

const file = readFileSync("test", "utf-8");
const valves = Object.fromEntries(
    file.split("\n").filter(l=>l.trim() != '').map(parseLine).map(v => [v.name, v]))

let ans1 = find_max_reward(30)
let ans2 = 0;
console.log(`${ans1},${ans2}`);
//console.log(min_distance("AA","BB"), 1)
//console.log(min_distance("GG","JJ"), 6)
